crates/biome_js_analyze/src/lint/style/no_parameter_assign.rs

```diff
@@ -64,48 +116,142 @@ declare_lint_rule! {
+            }
+
             if matches!(
                 declaration,
                 AnyJsBindingDeclaration::JsFormalParameter(_)
                     | AnyJsBindingDeclaration::JsRestParameter(_)
                     | AnyJsBindingDeclaration::JsArrowFunctionExpression(_)
                     | AnyJsBindingDeclaration::TsPropertyParameter(_)
             ) {
-                return binding.all_writes(ctx.model());
+                let param_reassignments: Vec<_> = binding
+                    .all_writes(model)
+                    .map(ProblemType::ParameterReassignment)
+                    .collect();
+
+                signals.extend(param_reassignments);
             }
         }
-        // Empty iterator that conforms to `AllBindingWriteReferencesIter` type.
-        std::iter::successors(None, |_| None)
+
+        signals
     }
 
-    fn diagnostic(ctx: &RuleContext<Self>, reference: &Self::State) -> Option<RuleDiagnostic> {
-        let param = ctx.query();
-        Some(
-            RuleDiagnostic::new(
-                rule_category!(),
-                reference.syntax().text_trimmed_range(),
-                markup! {
-                    "Reassigning a "<Emphasis>"function parameter"</Emphasis>" is confusing."
-                },
-            )
-            .detail(
-                param.syntax().text_trimmed_range(),
-                markup! {
-                    "The "<Emphasis>"parameter"</Emphasis>" is declared here:"
-                },
-            )
-            .note(markup! {
-                "Use a local variable instead."
-            }),
-        )
+    fn diagnostic(ctx: &RuleContext<Self>, state: &Self::State) -> Option<RuleDiagnostic> {
+        match state {
+            ProblemType::ParameterReassignment(expression) => Some(
+                RuleDiagnostic::new(
+                    rule_category!(),
+                    expression.syntax().text_trimmed_range(),
+                    markup! {
+                        "Reassigning a "<Emphasis>"function parameter"</Emphasis>" is confusing."
+                    },
+                )
+                    .detail(
+                        ctx.query().syntax().text_trimmed_range(),
+                        markup! {
+                        "The "<Emphasis>"parameter"</Emphasis>" is declared here:"
+                    },
+                    )
+                    .note(markup! {
+                    "Use a local variable instead."
+                }),
+            ),
+
+            ProblemType::PropertyAssignment(expression) => {
+                Some(
+                    RuleDiagnostic::new(
+                        rule_category!(),
+                        expression.syntax().text_trimmed_range(),
+                        markup! {
+                        "Assignment to a property of "<Emphasis>"function parameter"</Emphasis>" is confusing."
+                    })
+                        .note(markup! {"Use a local variable instead."}),
```

Comment:
I don't think this note is relevant for the property assignment case.
We could add something like:

```suggestion
                        .note(markup! {"Function callers usually don't expect the parameters they pass in to be modified. To avoid mutation, create a new instance and return it to the caller."}),
```